go-swagger added to test bench

aik-contrast opened the pull request:
This change adds support to run the test bench using the go-swagger framework. Up until now with the test bench we could only show vulnerabilities for the `gin` framework and the standard library. Recently the go agent has added new features which introduced support for the `go-swagger` framework. This PR aims to help with testing those features by adding an intentionally vulnerable go-swagger application to the test bench. This happens by adding a module inside the `/cmd` directory that can be used to make a new executable to create and run a swagger server which contains vulnerabilities. With this PR, the first vulnerability that we are adding is command injection through user supplied URL query.

Later on, adding further vulnerabilities can be done by attaching a new `go-swagger` handler of type `middleware.ResponderFunc` to handle the vulnerability and then use the server's `api.HTMLProducer` to visualize the given page. 

Routes, paths and the front-end aim to be the same as for gin and std and also the same template resources are used as for those two APIs.

mark-pictor-csec commented on cmd/go-swagger/restapi/operations/path_traversal/path_traversasl_front.go:
```diff
@@ -0,0 +1,56 @@
+// Code generated by go-swagger; DO NOT EDIT.
```
`traversal` is misspelled in the file name. Likewise for 3 other files.

aik-contrast replied:
corrected

aik-contrast commented:
A notable thing to point out with this PR is that OpenAPI v2.0 does not provide a way of directly adding user input through cookies. Since `go-swagger` is an implementation of OpenApi version 2.0, this means that the framework doesn't give expose a user facing entry point for data through cookies, which is why the routes for cookies have been removed from the route structure in the current PR. 
Reference: (https://swagger.io/docs/specification/2-0/describing-parameters/)

aik-contrast commented:
A notable thing to point out with this PR is that OpenAPI v2.0 does not provide a way of directly adding user input through cookies. Since `go-swagger` is an implementation of OpenApi version 2.0, this means that the framework doesn't expose a user facing entry point for cookie data, which is why the HTTP POST routes for cookies that we usually have for command injection have been removed from the current PR. 
Reference: (https://swagger.io/docs/specification/2-0/describing-parameters/)

mark-pictor-csec commented:
@contrast-jproberts noticed the root page didn't look right. Turns out the css is not being loaded. In the process of creating a patch for that, I found that embedded_spec.go (and other generated files) are not up-to-date with the yaml - running `go generate ./cmd/go-swagger/...` changed some descriptions.

mark-pictor-csec commented on pkg/serveswagger/html_producer.go:
```diff
@@ -0,0 +1,35 @@
+		}
+	}
+
+	t.ExecuteTemplate(w, "layout.gohtml", params)
```
error return value is not checked

mark-pictor-csec commented on pkg/servestd/servestd.go:
```diff
@@ -90,47 +87,12 @@ func newHandler(v common.Route) http.HandlerFunc {
 	}
 }
 
-func parseTemplates() error {
```
tests need updated for the new location of this file

mark-pictor-csec reviewed (changes requested):
app.css is not loaded, so the front page looks weird. [this discussion](https://github.com/go-swagger/go-swagger/issues/1375) may be helpful in getting the css served.

At least one test fails.

aik-contrast commented:
> @contrast-jproberts noticed the root page didn't look right. Turns out the css is not being loaded. In the process of creating a patch for that, I found that embedded_spec.go (and other generated files) are not up-to-date with the yaml - running `go generate ./cmd/go-swagger/...` changed some descriptions.

Yeah, the content in some of the generated files got changed because I wanted to try and reword the descriptions on the command injection endpoints to hopefully make them sound a little bit more accurate. Seemed like a small change, but I still should have reran the code generation just to be safe. Added and fixed. Currently looking into the missing `css` on the front page tho. Thanks for catching that 👍

mark-pictor-csec commented on cmd/go-swagger/restapi/configure_swagger_bench.go:
```diff
@@ -0,0 +1,137 @@
+
+func uiMiddleware(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if strings.Index(r.URL.Path, "/assets/") == 0 {
```
You should be able to use `if strings.HasPrefix(r.URL.Path, "/assets/") { /*....*/ }` here.

contrast-jproberts replied:
I also think it would be nice to save the handler as `var assetHandler` so the code here would be a bit simpler as `assetHandler.ServeHTTP(w,r)`. This is also a nit you can ignore.

mark-pictor-csec reviewed (approved):
One nit (feel free to ignore), otherwise looks good.

contrast-jproberts reviewed (approved):
LGTM. This task was much larger than any of us expected. Great job sticking with it 🎉

cixel reviewed (approved):
I'll echo what James said: nice work, and glad we're finally able to get this merged.
